### PR TITLE
Try to make ProcessNameMatchesScriptName test more reliable

### PR DESCRIPTION
**Pull request by adamsitnik**
This very simple test can sometimes fail: https://github.com/dotnet/runtime/issues/13757

I was not able to reproduce it, but based on the logs attached to the original issue I believe that the `GetTestFileName` might in theory be returning some non-ascii characters and Linux might not like it, so I've replaced it with `const` value
image

According to [man docs](https://man7.org/linux/man-pages/man5/proc.5.html) the file might be empty if the process is a zombie (it has already exited but parent process has not acknowledged that yet) so I've added a check for that.

> This read-only file holds the complete command line for the
              process, unless the process is a zombie.  In the latter case,
              there is nothing in this file: that is, a read on this file
              will return 0 characters.

I am not a Linux expert, but I guess that `sleep 600` could become a zombie right away only if there is a custom alias for the `sleep` command. I don't know if it's possible, but I think that we can just use full path to be 100% sure we are calling the actual `/bin/sleep`.

Fixes #13757

@tmds is there any chance you could take a look? It's nothing urgent, I just want to get rid of unstable tests from my areas.

### CODE REVIEW

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 1]
```diff
@@ -159,9 +159,9 @@ public void ProcessStart_UseShellExecute_OnUnix_SuccessWhenProgramInstalled(bool
         [PlatformSpecific(~TestPlatforms.OSX)] // On OSX, ProcessName returns the script interpreter.
         public void ProcessNameMatchesScriptName()
         {
-            string scriptName = GetTestFileName();
+            const string scriptName = nameof(ProcessNameMatchesScriptName);
```
Comment (tmds) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:162]: Unless the underlying filesystem imposes additional restrictions, Linux filenames can contain any bytes except `\0` and `/`.
.NET assumes the bytes are UTF-8.
So, I think the filename is not an issue. If it were, we should make sure `GetTestFileName` returns a valid filename on Linux instead of doing something specific in this test.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 2]
```diff
@@ -171,14 +171,26 @@ public void ProcessNameMatchesScriptName()
                     string stat = File.ReadAllText($"/proc/{process.Id}/stat");
                     Assert.Contains($"({scriptName.Substring(0, 15)})", stat);
                     string cmdline = File.ReadAllText($"/proc/{process.Id}/cmdline");
-                    Assert.Equal($"/bin/sh\0{filename}\0", cmdline);
 
-                    Assert.Equal(scriptName, process.ProcessName);
+                    // cmdLine can sometimes be null: https://github.com/dotnet/runtime/issues/13757
+                    // according to docs, it's when the process is a zombie: https://man7.org/linux/man-pages/man5/proc.5.html
```
Comment (tmds) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:176]: The `Assert` that fails was added to understand better why `ProcessName` doesn't end up with the proper value. It verifies some expectations of the .NET implementation.

Can we verify the failing case has a zombie process? We can check `/stat` for the `Z` flag.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 3]
```diff
@@ -171,14 +171,26 @@ public void ProcessNameMatchesScriptName()
...
+                    {
+                        Assert.Equal($"/bin/sh\0{filename}\0", cmdline);
+                        Assert.Equal(scriptName, process.ProcessName);
+                    }
                 }
                 finally
                 {
-                    process.Kill();
-                    process.WaitForExit();
+                    if (!process.HasExited)
```
Comment (tmds) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:189]: Why did you add this condition?

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 4]
```diff
@@ -159,9 +159,9 @@ public void ProcessStart_UseShellExecute_OnUnix_SuccessWhenProgramInstalled(bool
         [PlatformSpecific(~TestPlatforms.OSX)] // On OSX, ProcessName returns the script interpreter.
         public void ProcessNameMatchesScriptName()
         {
-            string scriptName = GetTestFileName();
+            const string scriptName = nameof(ProcessNameMatchesScriptName);
             string filename = Path.Combine(TestDirectory, scriptName);
-            File.WriteAllText(filename, $"#!/bin/sh\nsleep 600\n"); // sleep 10 min.
+            File.WriteAllText(filename, $"#!/bin/sh\n/bin/sleep 600\n"); // sleep 10 min.
```
Comment (tmds) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:164]: Is there a reason for changing the to explicitly using `/bin/sleep`? I don't think we care how the sleep is implemented.

Comment (adamsitnik) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:164]: > I am not a Linux expert, but I guess that `sleep 600` could become a zombie right away only if there is a custom alias for the `sleep` command. I don't know if it's possible, but I think that we can just use full path to be 100% sure we are calling the actual `/bin/sleep`.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 2 above)
Comment (adamsitnik) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:176]: Very good point, will do

Comment (tmds) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:176]: Thinking about it a bit more.

Probably this test fails when the `Assert` happens after the script has already exited. .NET will have reaped the process, and everything under `/proc` is gone.

This happens when sleep doesn't work, or when the asserts are delayed for 10 min. We can safely increase the 10 min timeout. And add an assert to check `!process.HasExited`.

**Comment by tmds** [comment]
Adam, thanks for working through flaky tests.
I think the timeout race (https://github.com/dotnet/runtime/pull/46292#discussion_r551176058) is the most likely cause for occasional failures of this test.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 2 above)
Comment (adamsitnik) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:176]: > This happens when sleep doesn't work

How would this be possible?

> when the asserts are delayed for 10 min

this is unlikely, as we just start the process and execute asserts right away

> everything under /proc is gone

Based on the CI logs it looks like not everything is gone. The previous line that asserts the content of `/proc/{process.Id}/stat` works just fine. Based on the available docs the only explanation I was able to find was the zombie process.

**Comment by jeffhandley** [comment]
@adamsitnik It looks like this PR is close to ready with just a little feedback left to address.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 4 above)
Comment (danmoseley) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:164]: I know on Alpine it's busybox, not an actual executable file, but it's symlinked to the same place so I guess that's OK:
```
LAPTOP-P6UJDVTA:~$ ls -l /bin/sleep
lrwxrwxrwx    1 root     root            12 Oct 21 09:23 /bin/sleep -> /bin/busybox
```

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:162]: I agree I do'nt see a problem with GetTestFileName(), and if it has a problem, it will be affecting other tests. Either way no change is needed here.

**Comment by danmoseley** [comment]
@adamsitnik looking at history I only see one failure
(@MattGal can you tell me why the query below does not show the failures reported in https://github.com/dotnet/runtime/issues/13757#issue-557983600 ? )

https://engsrvprod.kusto.windows.net/engineeringdata?query=H4sIAAAAAAAEAG2OQUvDQBCF74H8h7GXKERiUIqXePBQiFBRG%2fC83X022yS7YWZDEfzxZrWYi6dh3vc%2beA0kvEGmPgilyRcdvXXUWWcq6xyY3j13dcAg5N3fU5t%2fu09%2b%2f1Obb21i4dSCQRvrrLQw9FCR86fL69v1jbla%2bBah9YaqilYv7DVEntWArQq6hew02zHEYLUYv4OjkW2U7bOFPLJyuqWLmTA%2bpBinvi%2fu1uV9WQzgAzIqCtorQyNHaWR%2fhA60C4oDTH72c2o%2bR%2bTnZTm9TpgQN5QxE1GHGc6O7hpWGmmSJt%2bi3mERSAEAAA%3d%3d&web=0

TestResults 
| join kind=inner WorkItems on WorkItemId 
| join kind=inner Jobs on JobId
| where Finished >= now(-360d)
| where Method == "ProcessNameMatchesScriptName"
| where Result == 'Fail'
| where Branch != 'refs/pull/46181/merge' // bad pr
| project Started, Branch, Type, Method, QueueName1, Message, StackTrace

Started | Branch | Type | Method | QueueName1 | Message | StackTrace
-- | -- | -- | -- | -- | -- | --
2021-01-06 16:38:19.6450000 | refs/pull/46634/merge | System.Diagnostics.Tests.ProcessTests | ProcessNameMatchesScriptName | ubuntu.1804.armarch.open | Assert.Contains() Failure\nNot found: (ProcessNameMatc)\nIn value: 374 (dotnet) R 23 1 1 0 -1 256 1 0 0 0 0 0 0 0 20 0 1 0 87610 4096 1 18446744073709551615 0 0 0 0 0 0 0 4096 0 0 0 0 17 5 0 0 0 0 0 0 0 0 0 0 0 0 0\n | at System.Diagnostics.Tests.ProcessTests.ProcessNameMatchesScriptName() in /_/src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:line 172 at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) in /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs:line 378

This failure is in the first assert -- it's reading /proc/NNN/stat and finding 'dotnet' instead of 'ProcessNameMatchesScriptName_... something'. 

In the original bug, the assert was finding blank when reading /proc/NNN/cmdline.

I don't know exactly how the contents of these files change when the process exits, but I wonder whether these are consistent with it having exited, but not all been cleaned up yet. (The slightly surprising thing is we're not seeing eg FileNotFoundException, but then, there's only one failure in the history)

As you point out, it seems hard to imagine that 10 minutes could have elapsed. 

I suggest to just add `Assert(!process.HasExited)` just before the `Process.Kill()` and possibly `Assert(File.Exists("/bin/sleep"))` at the start. And call it fixed. If it's not, we will have better info. I don't think other changes are needed.

**Comment by MattGal** [comment]
> (@MattGal can you tell me why the query below does not show the failures reported in #13757 (comment) ? ) 

Taking a look.  We did have a volume issue last week hotfixed by doubling the capacity of machines ingesting to Kusto, which would have resulted in delayed ingestion.  Not sure this is that though.

**Comment by MattGal** [comment]
@danmoseley Actually, it's obvious no investigation needed.  The thing you linked is from 11/7/2019, 467 days ago.  We keep a bit more than 120 days of test results in Kusto, and even that costs a fair bit of money.  Once results age out they are naturally reclaimed in the data cluster.

**Comment by carlossanlop** [comment]
Ping @adamsitnik

**Comment by adamsitnik** [comment]
thank you all for great feedback, I am going to close the PR for now and get back to it when I have some free cycles
